### ISSUE TITLE
Incorrect jsonld expansion in presence of @vocab interacts with expansions of prefixes that use underscores

### ISSUE THREAD

**Issue by cmungall** [RDFLib/rdflib#2606]
Test case:

```json
{
  "classes": [
    {
      "p": "UBERON:0001062",
      "@type": "Term"
    }
  ],
   "@context": {
      "@vocab": "http://example.org/",
      "UBERON": {
         "@id": "http://purl.obolibrary.org/obo/UBERON_",
         "@prefix": true
      },
      "p": {
         "@type": "@id"
      }
   }
}
```

In the jsonld playground (https://json-ld.org/playground/), this gets parsed correctly with the compact URI `UBERON:0001062` being expanded to `<http://purl.obolibrary.org/obo/UBERON_0001062>`

_:b0 <http://example.org/classes> _:b1 .
_:b1 <http://example.org/p> <http://purl.obolibrary.org/obo/UBERON_0001062> .
_:b1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Term> .

However, when I convert via rdflib, the compact URI is not expanded, and is oddly downcased:

```turtle
@prefix : <http://example.org/> .

[] :classes [ a :Term ;
            :p <uberon:0001062> ] .
```

This is true of versions of rdflib from 6.1.0 through to 7.0.0

The correct behavior is rescued if:

1. A `/` is used rather than `_`
2. The `@vocab` element is removed

I am using this code to test:

```python
from rdflib import Graph

g = Graph()
g.parse(
    open("test.jsonld"),
    format="json-ld",
)
print(g.serialize(format="turtle"))
```

**Comment by pkalita-lbl** [comment]
I can confirm that I see the same behavior as @cmungall. One small correction though: when we see sporadic CI test failures that appear to be caused by this bug they happen on Ubuntu test runners with Python 3.8. 

Out of curiosity, I tried using the [python:3.8](https://hub.docker.com/layers/library/python/3.8/images/sha256-32280dd146dc84b459615d3b4c28597f5b299b9469cb9403a12abd245879a114?context=explore) and [python:3.9](https://hub.docker.com/layers/library/python/3.9/images/sha256-1fcc3e2d0128c39c20eb34e1a094c66f78cbea1de52428e0a5a82588bf0d50c7?context=explore) Docker images and got slightly different results with the above test script:

```text
# python:3.9
root@bfdb7dffbb75:/tmp# python test.py 
@prefix : <http://example.org/> .

[] :classes [ a :Term ;
            :p <uberon:0001062> ] .
```

```text
# python:3.8
root@e2fea7d79d01:/tmp# python test.py 
@prefix : <http://example.org/> .

[] :classes [ a :Term ;
            :p <file:///tmp/UBERON:0001062> ] .
```

**Comment by WhiteGobo** [comment]
It seems that the namespace `ns` in this `Context` file was set to be not a prefix(`ns.prefix == False`).

https://github.com/RDFLib/rdflib/blob/b0392f0a50db3b49cf6ec93bfbe0672a0fba16d6/rdflib/plugins/shared/jsonld/context.py#L354-L357

**Comment by WhiteGobo** [comment]
`_` is not part of `rdflib.plugins.shared.jsonld.context.URI_GEN_DELIMS`. So `http://purl.obolibrary.org/obo/UBERON_` wont be recognized as prefix.

Code where prefix is recognized:

https://github.com/RDFLib/rdflib/blob/b0392f0a50db3b49cf6ec93bfbe0672a0fba16d6/rdflib/plugins/shared/jsonld/context.py#L235-L238

**Comment by WhiteGobo** [comment]
https://github.com/RDFLib/rdflib/blob/b0392f0a50db3b49cf6ec93bfbe0672a0fba16d6/rdflib/plugins/shared/jsonld/context.py#L68

**Comment by WhiteGobo** [comment]
I would like to know if there is any reason to check for the ending letter if its a prefix or not.
Maybe its just better to get rid of this check entirely.

**Comment by cmungall** [comment]
Thanks for looking @WhiteGobo. It looks like that check is only triggered if the version is < 1.1. There was a difference in behavior between 1.0 and 1.1 regarding underscores, some background here:

https://github.com/w3c/json-ld-syntax/issues/329

this is the reason we have

```json
"UBERON": {
         "@id": "http://purl.obolibrary.org/obo/UBERON_",
         "@prefix": true
      },
```

and not the simpler


```json
"UBERON": "http://purl.obolibrary.org/obo/UBERON_"
```

I'm not very familiar with the rdflib internals here and I feel that some kind of logic regarding the change of behavior between 1.0 and 1.1 may be important here

**Comment by WhiteGobo** [comment]
Oh i didnt really understood the role of `@prefix` in the code :) thx

It seems it was planned that you can explicitly set `"_"` as working ending character. Can you set `URI_GEN_DELIMS` beforehand?

```
from rdflib.plugins.shared.jsonld import context
context.URI_GEN_DELIMS = (*context.URI_GEN_DELIMS, "_")
```

They mentioned this in one of their meetings. But if this was intended, a list would be the expected type of `URI_GEN_DELIMS`

**Comment by WhiteGobo** [comment]
As far as i understand, the discussions in this link [w3c/json-ld-syntax/issues/329](https://github.com/w3c/json-ld-syntax/issues/329), the use `@prefix` breaks sometimes some other code. But there isnt any explanation, why it is generally disabled for 1.1

They have explicitly decided not to include "_" in `URI_GEN_DELIMS`. So i will take back my PR.

I would consider making `URI_GEN_DELIMS` a list, so one can add `"_"` or any other character if needed or enabling `@prefix` for 1.1

**Comment by niklasl** [comment]
As you note, `@prefix` is needed for any term intended to be used as a prefix if its IRI doesn't end in a IRI-delimiting character.

This is mentioned in the [spec](https://www.w3.org/TR/json-ld11/#compact-iris) which references [the announcement](https://lists.w3.org/Archives/Public/public-rdf-comments/2018Jan/0002.html), in turn linking to https://github.com/json-ld/json-ld.org/pull/511, explained in more detail in https://github.com/json-ld/json-ld.org/issues/469.

In short, this is to avoid declaring terms which end up being expanded as prefixes unintentionally, such as:
> When you define "geo" as a term in a context, you clobber the absolute geo: URI scheme.

If you mutate URI_GEN_DELIMS you'll end up creating a deviating implementation, which may exasperate this problem in the wild. Please update your context to use `@prefix`  if you use non-standard namespace forms, to ensure interoperability.

**Comment by WhiteGobo** [comment]
Mh some things to clarify for me. Im not sure if I understand the difference between **simple term definition** and **expanded term definition**.
If i understand correct, **simple term definitions** should produce not expandind prefixes, eg:
```
{
  "classes": [
    {
      "p": "UBERON:0001062",
      "@type": "Term"
    }
  ],
   "@context": {
      "@vocab": "http://example.org/",
      "UBERON": "http://purl.obolibrary.org/obo/UBERON_",
      "p": {"@type": "@id"}
   }
}
```
will result in `<uberon:0001062>`

But the **expanded term definition** if `@prefix` is set should produce an expanding prefix, eg:
```
{
  "classes": [
    {
      "p": "UBERON:0001062",
      "@type": "Term"
    }
  ],
   "@context": {
      "@vocab": "http://example.org/",
      "UBERON": {
         "@id": "http://purl.obolibrary.org/obo/UBERON_",
         "@prefix": true
      },
      "p": {"@type": "@id" }
   }
}
```
will result in `<http://purl.obolibrary.org/obo/UBERON_0001062>`?

**Comment by niklasl** [comment]
@WhiteGobo Yes, that is correct. This is what e.g. https://json-ld.org/playground/ and  https://github.com/niklasl/trld do (both compliant with the 1.1 algorithm).

**Comment by cmungall** [comment]
There was a lot of discussion here, but to summarize, we are agreed that this is a bug in rdflib?

**Comment by WhiteGobo** [comment]
As far as i remember, yes it was. But i cant remember the exact reason and we havent talked about it that explicitly.
Also because nobody has talked about this, you can use version `1.1`. This problem exists only for version `1.0`

```python
g = Graph().parse(open("test.jsonld"), format="json-ld", version=1.1)
```

**Comment by WhiteGobo** [comment]
Ehm no i dont think this a bug. As mentioned  [here](https://github.com/w3c/json-ld-syntax/issues/329#issuecomment-587130578) that kind of prefix wont work on version `1.0`.

**Comment by cmungall** [comment]
Ah, I see, I do in fact get the desired behavior when I set `version="1.1"`.

I think the behavior under 1.0 is still odd. When I try it in the jsonld playground in 1.0 mode, it throws an error due to `@prefix`. It may be reasonable to assume that rdflib is ignoring constructs it doesn't recognize in 1.0 mode - however, when I try this in the playground:

```json
{
  "classes": [
    {
      "p": "UBERON:0001062",
      "@type": "Term"
    }
  ],
   "@context": {
      "@vocab": "http://example.org/",
      "UBERON": {
         "@id": "http://purl.obolibrary.org/obo/UBERON_"
      },
      "p": {
         "@type": "@id"
      }
   }
}
```

I get back the correctly expanded URI:

```turtle
_:b0 <http://example.org/classes> _:b1 .
_:b1 <http://example.org/p> <http://purl.obolibrary.org/obo/UBERON_0001062> .
_:b1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Term> .
```
